pteranodon/autoconnect.py

```diff
@@ -0,0 +1,234 @@
+import asyncio
+from typing import List, Optional, Tuple, Any, Dict
+import os
+import re
+import socket
+
+from pymavlink import mavutil
+
+
+# TODO_: Better name?
+class AutoConnect:
+    """
+    Searches serial devices and, if a server ip address is provided, that server for open MAV SDK servers.
+    """
+
+    def __init__(self, server_ip_addr: Optional[str]):
+        """
+        Searches serial devices and, if a server ip address is provided, that server for open MAV SDK servers.
+
+        :param server_ip_addr: The server IP address to search for MAV SDK searches, None if only serial devices should be searched.
+        :type server_ip_addr: Optional[str]
+        """
+        self._server_addr = server_ip_addr
+        self._should_check_server = self._server_addr is not None
+
+    async def _test_port_open(
+        self,
+        port: int,
+        socket_kind: int,
+        timeout: float = 1.0,
+    ) -> Optional[Tuple[int, bool]]:
+        """
+        Attempts to connect to a port with a specified timeout and socket kind to determine if the port is open.
+
+        :param port: The port to check
+        :type port: int
+        :param socket_kind: The type of socket to open (SOCK_STREAM for TCP, SOCK_DGRAM for UDP)
+        :type socket_kind: socket.SocketKind
+        :param timeout: The timeout for the connect call
+        :type timeout: float
+        :return: The port tested and whether the port is open, None if invalid parameters
+        :rtype: Optional[Tuple[int, bool]]
+        """
+        # Must have a server address
+        if not self._should_check_server:
+            return None
+
+        # Only supporting TCP and UDP
+        if socket_kind not in [socket.SOCK_STREAM, socket.SOCK_DGRAM]:
+            return None
+
+        with socket.socket(socket.AF_INET, socket_kind) as sock:
+            sock.settimeout(timeout)
+            try:
+                sock.connect((self._server_addr, port))
+            except TimeoutError:
+                return port, False
+
+        return port, True
+
+    def fetch_possible_serial_devs(self) -> List[str]:
```

Comment:
This can honestly be a private static method which we can access with either a property or cached_property. I saw your comment mentioned pylint requires at least two public methods, but those can just be like search_possible_servers, get_possible_servers, something like that. The less stuff exposed in this the better I think. Will leave another comment talking about integration which could make more sense then

Reply:
So I just added some caching with the ports. I don't think caching properties is the best move here because just because we have loaded it already doesn't mean it won't ever change, especially with serial. Since serial devices are just filtering a list and are not async calls, I cannot see caching the serial devs being that helpful. I did add the possible option to use the cached TCP or UDP ports instead of searching for the open ones again since that may be more redundant if run multiple times.